### PR TITLE
Add fast path for wildcard patterns that contains no wildcard characters

### PR DESCRIPTION
**Pull request by daxian-dbw**
# PR Summary

At many places in the code base, PowerShell uses `WildcardPattern` for strings that don't really contain any wildcard characters. For example, `Import-Module` use `WildcardPattern` for entries of `ExposedFunctions`, `ExposedCmdlets`, `ExposedAlias` and etc. even if those entries contain no wildcard characters. In that case, `WildcardPattern.IsMatch` essentially should just do a string comparison, but we are go through the same code path that handles real wildcard patterns, while results in unnecessary cycles and allocations.

This PR adds a fast path for `WildcardPattern.IsMatch(string)` for that case -- simply do a string comparison when the pattern string contains no special characters (wildcard characters `*`, `?`, `[`, `]` and the escape character backtick `).

The following is the the memory allocation when the module analysis is triggered.
As is shown, `int32[]` allocation drops drastically, and also the type `PatternPositionVisitor` is completely avoided.
This is because the pattern matching happens in module analysis (e.g. `ModuleIntrinsics.ExportModuleMembers`) are now going through the fast path.

image

Also, `WildcardPattern.IsMatch` is about 2x faster in steady state when the pattern string contains no wildcard characters:

#### Before the change

```
PS:1> $p = [WildcardPattern]::new("abcdefghi")
PS:18> Measure-Command { foreach ($i in 1..100000) { $p.IsMatch("abc") } }

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 103
Ticks             : 1030111
TotalDays         : 1.19225810185185E-06
TotalHours        : 2.86141944444444E-05
TotalMinutes      : 0.00171685166666667
TotalSeconds      : 0.1030111
TotalMilliseconds : 103.0111
```

#### After the change
```
PS:1> $p = [WildcardPattern]::new("abcdefghi")
PS:31> Measure-Command { foreach ($i in 1..100000) { $p.IsMatch("abc") } }

Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 59
Ticks             : 594633
TotalDays         : 6.88232638888889E-07
TotalHours        : 1.65175833333333E-05
TotalMinutes      : 0.000991055
TotalSeconds      : 0.0594633
TotalMilliseconds : 59.4633
```

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs [hunk 1]
```diff
@@ -1039,11 +1039,7 @@ internal static Collection<PSSnapInInfo> ReadEnginePSSnapIns()
 
                 string moduleName = Path.Combine(applicationBase, defaultMshSnapinInfo.AssemblyName + ".dll");
 
-                if (File.Exists(moduleName))
-                {
-                    moduleName = Path.Combine(applicationBase, defaultMshSnapinInfo.AssemblyName + ".dll");
-                }
-                else
+                if (!File.Exists(moduleName))
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]: Is this change related?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs (hunk 1 above)
Comment (adityapatwardhan) [reply src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]: Also, do we have any wildcard tests?

File: src/System.Management.Automation/engine/regex.cs [hunk 2]
```diff
@@ -148,7 +146,26 @@ private void Init()
...
                 if (Pattern.Length == 1 && Pattern[0] == '*')
                 {
-                    _isMatch = _ => true;
+                    _isMatch = s_matchAll;
+                }
+                else if (Pattern.IndexOfAny(s_specialChars) == -1)
+                {
+                    // No special characters present in the pattern, so we can just do a string comparison.
+                    StringComparison stringComparison;
+                    if (Options.HasFlag(WildcardOptions.IgnoreCase))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/regex.cs]: We could unroll the if-s to switch for better readability.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

File: src/System.Management.Automation/engine/regex.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/regex.cs]: The `WildcardOptions` enum has `[Flags]` declared, so `Options` could be the combination of multiple `WildcardOptions` members. That makes it not friendly to `switch`. For example, the `Compile` member is completely ignored in this scenario, though it could be declared in `Options`.

File: src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs (hunk 1 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]: This change is just a quick cleanup of something that I ran into :) It does `moduleName = Path.Combine(applicationBase, defaultMshSnapinInfo.AssemblyName + ".dll");` twice unnecessarily.

Comment (daxian-dbw) [reply src/System.Management.Automation/singleshell/config/MshSnapinInfo.cs]: As for tests, I don't think we have any tests explicitly for `WildcardPattern`, but it's been indirectly exercised by all other tests.

File: src/System.Management.Automation/engine/regex.cs (hunk 2 above)
Comment (rjmholt) [inline src/System.Management.Automation/engine/regex.cs]: Given the depth of this `if` statement and to @iSazonov's point about readability, I think this method is now better defined as:

```csharp
private void Init()
{
    if (_isMatch != null) { return; }

    if (Pattern.Length == 1 && Pattern[0] == '*')
    {
        _isMatch = s_matchAll;
        return;
    }

    if (Pattern.IndexOfAny(s_specialChars) == -1)
    {
        // String comparison definition
        _isMatch = str => string.Equals(str, Pattern, stringComparison);
       return;
    }

    _isMatch = new WildcardPatternMatcher(this).IsMatch;
}
```

Really though, `Init()`should just be a `Lazy<Predicate<string>>` private property

**Review by rjmholt — COMMENTED** [review]
I think some wildcard tests would be good -- maybe not all in this PR, but we should create the precedent by creating a place for them and adding one or two maybe

File: src/System.Management.Automation/engine/regex.cs (hunk 2 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/regex.cs]: This does look better. Will make the change.

File: src/System.Management.Automation/engine/regex.cs [hunk 3]
```diff
@@ -49,24 +49,22 @@ public enum WildcardOptions
     /// </summary>
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @daxian-dbw, your last commit had 38 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

SemanticVersion api tests.Error handling.Missing parts: 'null': '$null'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: 'NullString': ''
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: 'EmptyString': '<empty>'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: '-'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: '.'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\Basic\SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @daxian-dbw, your last commit had 38 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

SemanticVersion api tests.Error handling.Missing parts: 'null': '$null'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: 'NullString': ''
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: 'EmptyString': '<empty>'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: '-'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: '.'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/regex.cs]: @daxian-dbw, your last commit had 38 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

SemanticVersion api tests.Error handling.Missing parts: 'null': '$null'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: 'NullString': ''
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: 'EmptyString': '<empty>'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: '-'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```
SemanticVersion api tests.Error handling.Missing parts: '.'
```powershell
Method invocation failed because [System.Management.Automation.LanguagePrimitives] does not contain a method named 'IsNull'.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/engine/Basic/SemanticVersion.Tests.ps1: line 207
207:             if ([LanguagePrimitives]::IsNull($version)) {

```

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@daxian-dbw Please have a look at the test failures.

**Comment by daxian-dbw** [comment]
@adityapatwardhan The failures might be caused by conflicts with the merging of `DBNull.Value` PR. I just rebased.

**Review by rjmholt — APPROVED** [review]

**Comment by daxian-dbw** [comment]
@adityapatwardhan The issue reported by Codacy is not a real issue.

File: src/System.Management.Automation/engine/regex.cs [hunk 4]
```diff
@@ -50,24 +50,22 @@ public enum WildcardOptions
...
         private const char escapeChar = '`';
 
-        //
         // we convert a wildcard pattern to a predicate
-        //
         private Predicate<string> _isMatch;
 
-        //
+        // chars that are considered special in a wildcard pattern
+        private static readonly char[] s_specialChars = new char[] { '*', '?', '[', ']', '`' };
```
Comment (vexx32) [inline src/System.Management.Automation/engine/regex.cs]:
```suggestion
        private static readonly char[] s_specialChars = new[] { '*', '?', '[', ']', '`' };
```

**Review by vexx32 — COMMENTED** [review]
I think this is what Codacy is talking about 😊

File: src/System.Management.Automation/engine/regex.cs (hunk 4 above)
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/regex.cs]: Yes. I personally don't have preference to `new[]` over `new char[]`. But I guess I will just go with it to mute Codacy :)

Comment (vexx32) [reply src/System.Management.Automation/engine/regex.cs]: Something something redundant declarations. Abbreviations are becoming all the rage. :joy:

**Comment by daxian-dbw** [comment]
@adityapatwardhan Gentle ping :) I think this PR is ready to merge.

**Comment by iSazonov** [comment]
Not for the PR. We could add special case for patterns like "test*" - translated in StartsWith().
